### PR TITLE
[luajit] Initial integration

### PR DESCRIPTION
**Pull request by ligurio**
Based on Lua project [1], uses LuaJIT version v2.1.

1. https://github.com/ispras/oss-sydr-fuzz/tree/master/projects/lua

### CODE REVIEW

File: projects/luajit/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+	sydr-fuzz -c luaL_loadstring.toml cmin
+
+Collect and report coverage:
+
+    # sydr-fuzz -c luaL_loadstring.toml cov-report
+
+Building HTML report:
+
+	$ sydr-fuzz -c luaL_loadstring.toml cov-show -- -format=html > index.html
```
Comment (kobrineli) [inline projects/luajit/README.md:40]: Change TAB to spaces and change `$` to `#`.

File: projects/luajit/build.sh [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+: ${LD:="${CXX}"}
+: ${LDFLAGS:="${CXXFLAGS}"}  # to make sure we link with sanitizer runtime
+
+cmake_args=(
+    -DUSE_LUAJIT=ON
+    -DLUA_VERSION=224129a8e64bfa219d35cd03055bf03952f167f6
+    -DOSS_FUZZ=OFF
+    -DENABLE_ASAN=OFF
+    -DENABLE_UBSAN=OFF
```
Comment (kobrineli) [inline projects/luajit/build.sh:36]: Is it ok that sanitizers are turned off?

File: projects/luajit/build.sh [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+CXX=afl-clang-fast++
+CFLAGS="-g -fsanitize=fuzzer"
+CXXFLAGS="-g -fsanitize=fuzzer"
+
+cmake_args=(
+    -DUSE_LUAJIT=ON
+    -DLUA_VERSION=224129a8e64bfa219d35cd03055bf03952f167f6
+    -DOSS_FUZZ=OFF
+    -DENABLE_ASAN=OFF
+    -DENABLE_UBSAN=OFF
```
Comment (kobrineli) [inline projects/luajit/build.sh:82]: Should the sanitizers be turned off? Here and in AFL++ cmplog build.

**Review by kobrineli — CHANGES_REQUESTED** [review]
Hi! Thanks for the PR.

File: projects/luajit/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    # cd /fuzz
+
+Run hybrid fuzzing:
+
+    # sydr-fuzz -c luaL_loadstring.toml run
+
+Corpus minimization (step is required for AFL++):
+
+	sydr-fuzz -c luaL_loadstring.toml cmin
```
[the remark this answers is not in the document]
Comment (ligurio) [reply projects/luajit/README.md:32]: Fixed.

File: projects/luajit/README.md (hunk 1 above)
Comment (ligurio) [reply projects/luajit/README.md]: Fixed here and above.

File: projects/luajit/build.sh (hunk 2 above)
Comment (ligurio) [reply projects/luajit/build.sh:36]: Yes, vanilla LuaJIT doesn't support sanitizers.

File: projects/luajit/build.sh (hunk 3 above)
Comment (ligurio) [reply projects/luajit/build.sh:82]: Yes.

File: projects/luajit/build.sh [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  echo "Copying for $module";
+  cp $f /
+  [[ -e $corpus_dir ]] && cp -r $corpus_dir /corpus_$module
+  [[ -e $corpus_dir.dict ]] && cp $corpus_dir.dict /$module.dict
+done
+
+# Build the project for AFL++.
+CC=afl-clang-fast
+CXX=afl-clang-fast++
+CFLAGS="-g -fsanitize=fuzzer"
```
[the remark this answers is not in the document]
Comment (ligurio) [reply projects/luajit/build.sh:74]: Removed.

```diff
--- a/projects/luajit/build.sh
+++ b/projects/luajit/build.sh
@@ -71,8 +71,8 @@ done
 # Build the project for AFL++.
 CC=afl-clang-fast
 CXX=afl-clang-fast++
-CFLAGS="-g -fsanitize=fuzzer"
-CXXFLAGS="-g -fsanitize=fuzzer"
+CFLAGS="-g"
+CXXFLAGS="-g"
 
 cmake_args=(
     -DUSE_LUAJIT=ON
```

**Comment by ligurio** [comment]
> Tried to remove DOWNLOAD_EXTRACT_TIMESTAMP parameter from BuildLuaJIT.cmake in lua-c-api-tests, this helped.

`DOWNLOAD_EXTRACT_TIMESTAMP` is available in CMake since 3.24. [^1]

> Why do we need DOWNLOAD_EXTRACT_TIMESTAMP for luajit?

Seems it is applicable when `URL` download method in `ExternalProject_Add` and could be removed. I'll fix it.

[^1]: https://cmake.org/cmake/help/latest/module/ExternalProject.html#commands
[^2]: https://cmake.org/cmake/help/latest/policy/CMP0135.html#policy:CMP0135

**Comment by SweetVishnya** [comment]
I'll update cmake to latest version in base docker image.

File: projects/luajit/build.sh [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+  [[ -e $corpus_dir ]] && cp -r $corpus_dir /corpus_$module
+  [[ -e $corpus_dir.dict ]] && cp $corpus_dir.dict /$module.dict
+done
+
+# Build the project for AFL++.
+CC=afl-clang-fast
+CXX=afl-clang-fast++
+CFLAGS="-g"
+CXXFLAGS="-g"
+
```
Comment (kobrineli) [inline projects/luajit/build.sh:76]: Maybe here must be these lines after flags setting:

```
export AFL_LLVM_DICT2FILE=/afl++.dict
export AFL_LLVM_DICT2FILE_NO_MAIN=1
```

Without them the `afl++.dict` is not created and thus the fuzzing crashes.

**Review by kobrineli — CHANGES_REQUESTED** [review]

File: projects/luajit/build.sh (hunk 6 above)
Comment (ligurio) [reply projects/luajit/build.sh:76]: Updated:

```diff
--- a/projects/luajit/build.sh
+++ b/projects/luajit/build.sh
@@ -74,6 +74,9 @@ CXX=afl-clang-fast++
 CFLAGS="-g"
 CXXFLAGS="-g"
 
+export AFL_LLVM_DICT2FILE=/afl++.dict
+export AFL_LLVM_DICT2FILE_NO_MAIN=1
+
 cmake_args=(
     -DUSE_LUAJIT=ON
     -DLUA_VERSION=224129a8e64bfa219d35cd03055bf03952f167f6
```

File: projects/luajit/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+################################################################################
+
+FROM sydr/ubuntu20.04-sydr-fuzz
+
+MAINTAINER Sergey Bronnikov
+
+# Install build dependencies.
+RUN apt-get update && apt-get install -y \
+	build-essential ninja-build cmake make \
+	libreadline-dev libunwind-dev
```
Comment (kobrineli) [inline projects/luajit/Dockerfile:24]: Please, remove all the dependencies except of `libunwind-dev` and `libreadline-dev`.
They are installed in the base sydr-fuzz docker container.

**Review by kobrineli — CHANGES_REQUESTED** [review]

File: projects/luajit/Dockerfile (hunk 7 above)
Comment (kobrineli) [reply projects/luajit/Dockerfile:23]: @ligurio 
Base image dockerfile can be found in `oss-sydr-fuzz/docker/ubuntu20.04-sydr-fuzz/`

**Review by kobrineli — APPROVED** [review]
Thanks for the fixes!

**Pull request merged**